File: js/015. 3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort(compare);
    var ans=new Array();
    console.log(nums)
    for (var i = 0; i < nums.length; i++) {
    	if(i==0||nums[i]!=nums[i-1]){
    		var s=i+1
	    	var e=nums.length-1;
	    	while(s<e){
	    		var t=nums[i]+nums[s]+nums[e];
	    		if (t==0) {
	    			var a=[nums[i],nums[s],nums[e]];
	    			ans.push(a);
	    			while(nums[s+1]==nums[s]){
	    				s+=1
	    			}
	    			s+=1;
	    			e-=1;
	    		}else{
	    			if(t>0){
	    				e-=1;
	    			}else{
	    				s+=1;
	    			}
	    		}
	    	}
    	}
    }
    return ans
};
var compare = function(a,b){
	return a-b; 
}

File: ruby/015. 3Sum.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
    nums=nums.sort
    length=nums.length
    ans=[]
    for i in 0..length-3
        if i==0 || nums[i]!=nums[i-1]
            s,e=i+1 , length-1
            while s<e
                test=nums[i]+nums[s]+nums[e]
                if test==0
                    ans.push([nums[i],nums[s],nums[e]])
                end
                if test<=0
                    while nums[s]==nums[s+1]
                        s+=1
                    end
                    s+=1
                end
                if test>=0
                    while nums[e]==nums[e-1]
                        e-=1
                    end
                    e-=1
                end
            end
        end
    end
    return ans
end
